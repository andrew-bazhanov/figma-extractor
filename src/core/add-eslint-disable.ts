const maxLineRuleComment = '/* eslint-disable max-lines */';

const namingConventionRuleComment = '/* eslint-disable @typescript-eslint/naming-convention */';

export const addEslintDisableAtTheTopOfText = (
  content: string,
  options: {
    disabledMaxLines: boolean;
    disabledTypescriptNamingConvention: boolean;
  },
) => {
  return `
    ${options.disabledMaxLines ? maxLineRuleComment : ''}
    ${options.disabledTypescriptNamingConvention ? namingConventionRuleComment : ''}
    ${content}
    `;
};
