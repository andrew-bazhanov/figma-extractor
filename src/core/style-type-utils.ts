import { StyleType } from 'figma-js';

// eslint-disable-next-line @typescript-eslint/naming-convention
type StyleTypePredicate = (item: { style_type: StyleType }) => boolean;

const getStyleTypePredicate = (styleType: StyleType): StyleTypePredicate => {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  return ({ style_type }) => style_type === styleType;
};

export const styleTypeUtils = {
  isFill: getStyleTypePredicate('FILL'),
  isEffect: getStyleTypePredicate('EFFECT'),
  isText: getStyleTypePredicate('TEXT'),
};
