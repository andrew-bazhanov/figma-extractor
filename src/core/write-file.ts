import fs from 'fs';
import path from 'path';

const fileLengthRuleDisableComment = '/* eslint-disable max-lines */';

const namingConventionRuleDisableComment =
  '/* eslint-disable @typescript-eslint/naming-convention */';

const defaultOptions = {
  useEslintDisabledRules: true,
};

// NOTE: temporarily changed writeFile to writeFileSync to make testing easier:
// it is unclear how to test callback version without introducing promises.
// But even with promises, the logic seems to "just write" contents without returning the promise itself,
// which is confusing a bit, because it is impossible to handle promise fulfillment.
export const writeFile = (
  template: string,
  filePath: string,
  options?: {
    useEslintDisabledRules?: boolean;
  },
) => {
  const dir = path.dirname(filePath);
  const useEslintDisabledRules = { ...defaultOptions, ...options }.useEslintDisabledRules;

  // Create directory in advance if there is no one
  if (dir && !fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }

  const content = useEslintDisabledRules
    ? `
    ${fileLengthRuleDisableComment}
    ${namingConventionRuleDisableComment}
    ${template}
    `
    : template;

  try {
    fs.writeFileSync(filePath, content);
    console.log(`Generate ${filePath} is completed`);
  } catch {
    console.log(`Formatting ${filePath} file...`);
  }
};
