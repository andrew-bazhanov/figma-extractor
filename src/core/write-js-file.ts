import fs from 'fs';
import path from 'path';

const fileLengthRuleDisableComment = '/* eslint-disable max-lines */';

const namingConventionRuleDisableComment =
  '/* eslint-disable @typescript-eslint/naming-convention */';

// NOTE: temporarily changed writeFile to writeFileSync to make testing easier:
// it is unclear how to test callback version without introducing promises.
// But even with promises, the logic seems to "just write" contents without returning the promise itself,
// which is confusing a bit, because it is impossible to handle promise fulfillment.
export const writeJsFile = (template: string, filePath: string) => {
  const dir = path.dirname(filePath);

  // Create directory in advance if there is no one
  if (dir && !fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }

  try {
    fs.writeFileSync(
      filePath,
      `
      ${fileLengthRuleDisableComment}
      ${namingConventionRuleDisableComment}
      ${template}
      `,
    );
    console.log(`Generate ${filePath} is completed`);
  } catch {
    console.log(`Formatting ${filePath} file...`);
  }
};
