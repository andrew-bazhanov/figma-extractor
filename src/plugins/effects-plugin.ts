import path from 'path';

import { getEffectName } from '../lib/color/get-effect-name/get-effect-name';
import { getEffectStyles } from '../lib/get-effect-styles';
import { stringifyRecordsWithSort } from '../lib/stringify';
import { Plugin } from './types';

export const effectsPlugin: Plugin = (
  { config, styleTypeUtils, writeFile, runFormattingFile },
  { styleMetadata, fileNodes },
) => {
  if (!config?.styles?.effects?.disabled) {
    const metaEffects = styleMetadata.filter(styleTypeUtils.isEffect);
    const effects = getEffectStyles(
      metaEffects,
      fileNodes,
      config?.styles?.effects?.keyName ?? getEffectName,
    );

    const effectTemplate = `module.exports = {boxShadow: ${stringifyRecordsWithSort(effects)}};`;
    writeFile(effectTemplate, path.join(config?.styles?.exportPath || '', 'effects.js'));
    runFormattingFile(path.join(config?.styles?.exportPath || '', 'effects.js'));
  }
};

effectsPlugin.pluginName = 'effects';
