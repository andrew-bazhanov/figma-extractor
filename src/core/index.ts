import { ClientInterface, FileNodesResponse, FullStyleMetadata } from 'figma-js';

import { Plugin } from '../plugins';
import { runFormattingFile } from './run-formatting-file';
import { styleTypeUtils } from './style-type-utils';
import { writeJsFile } from './write-js-file';

export interface FigmaData {
  figmaClient: ClientInterface;
  styleMetadata: readonly FullStyleMetadata[];
  fileNodes: FileNodesResponse;
}

export interface Core {
  config: Config;
  rootPath: string;
  plugins: Plugin[];
  styleTypeUtils: typeof styleTypeUtils;
  writeJsFile: typeof writeJsFile;
  runFormattingFile: typeof runFormattingFile;
}

export const createCore = (args: { config: Config; rootPath: string; plugins: Plugin[] }): Core => {
  return {
    ...args,
    styleTypeUtils,
    writeJsFile,
    runFormattingFile: runFormattingFile,
  };
};
