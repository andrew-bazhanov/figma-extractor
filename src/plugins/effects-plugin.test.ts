import { createCore } from '../core';
import { fileNodes } from '../utils/generate-styles/fixtures/file-nodes';
import { styleMetadata } from '../utils/generate-styles/fixtures/style-metadata';
import { effectsPlugin } from './effects-plugin';

describe('effectsPlugin', () => {
  it('should write js file and run formatting tool', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          effects: {},
        },
      },
      plugins: [],
      rootPath: '/root-path',
    });

    core.writeFile = jest.fn();

    effectsPlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any);

    expect(core.writeFile).toHaveBeenCalled();
    expect((core.writeFile as jest.Mock).mock.calls[0][1]).toBe('/export-path/effects.js');
  });

  it('should write generated date for default config', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          effects: {},
        },
      },
      plugins: [],
      rootPath: '/root-path',
    });

    core.writeFile = jest.fn();

    effectsPlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any);

    expect(core.writeFile).toHaveBeenCalled();
    expect((core.writeFile as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
  });

  it('should be skipped due to disabled field', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          effects: {
            disabled: true,
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
    });

    core.writeFile = jest.fn();

    effectsPlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any);

    expect(core.writeFile).not.toHaveBeenCalled();
  });

  it('should create data with custom keyName function', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          effects: {
            keyName: (name?: string) => name + '__extra',
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
    });

    core.writeFile = jest.fn();

    effectsPlugin(core, { styleMetadata: styleMetadata.styles, fileNodes } as any);

    expect(core.writeFile).toHaveBeenCalled();
    expect((core.writeFile as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
  });
});
