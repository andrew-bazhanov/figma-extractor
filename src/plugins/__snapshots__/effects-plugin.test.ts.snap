// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`effectsPlugin should create data with custom keyName function 1`] = `
"
    /* eslint-disable max-lines */
    
    module.exports = {boxShadow: {"colorful/shadow300__extra":"0px 10px 46px rgba(51, 229, 101, 0.70)","dark/shadow300__extra":"0px 10px 46px rgba(0, 0, 0, 0.50)","dark/shadow400__extra":"0px 0px 10px rgba(178, 178, 0, 1.00)","light/shadow300__extra":"0px 10px 46px rgba(255, 255, 255, 0.50)","monochrome/shadow300__extra":"0px 10px 46px rgba(142, 142, 142, 0.50)","shadow200__extra":"0px 18px 88px rgba(24, 39, 75, 0.14)","shadow300__extra":"0px 10px 46px rgba(24, 24, 24, 0.50)"}};
    "
`;

exports[`effectsPlugin should write generated date for default config 1`] = `
"
    /* eslint-disable max-lines */
    
    module.exports = {boxShadow: {"colorful/shadow300":"0px 10px 46px rgba(51, 229, 101, 0.70)","dark/shadow300":"0px 10px 46px rgba(0, 0, 0, 0.50)","dark/shadow400":"0px 0px 10px rgba(178, 178, 0, 1.00)","light/shadow300":"0px 10px 46px rgba(255, 255, 255, 0.50)","monochrome/shadow300":"0px 10px 46px rgba(142, 142, 142, 0.50)","shadow200":"0px 18px 88px rgba(24, 39, 75, 0.14)","shadow300":"0px 10px 46px rgba(24, 24, 24, 0.50)"}};
    "
`;
